### PR TITLE
Group Manager with backend and UI

### PR DESCRIPTION
**Pull request by puneethreddy20**

### CODE REVIEW

File: templates/admins_sidebar.html [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{{ define "sidebar" }}
+
+<nav class="w3-sidebar w3-collapse w3-white w3-animate-left" style="z-index:3;width:300px;" id="mySidebar"><br>
+    <div class="w3-container w3-row">
+        <div class="w3-col s4">
+            <img src="https://v-groupmanager-dev-001.us-west-2.aws.symcpe.net:443/images/avatar2.png" class="w3-circle w3-margin-right" style="width:46px">
```
Comment (cviecco) [inline templates/admins_sidebar.html]: all links should be relative

**Comment by puneethreddy20** [comment]
made the required changes & seperated the code into multiple files.

File: api.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+//Display all users in Target LDAP --required
+func (state *RuntimeState) GetallUsers_Handler(w http.ResponseWriter, r *http.Request) {
+	var AllUsers_TargetLdap GetUsers
+
+	AllUsers, err := state.GetallUsers(state.Config.TargetLDAP.UserSearchBaseDNs, state.Config.TargetLDAP.UserSearchFilter, []string{"uid"})
+
+	if err != nil {
+		log.Println(err)
```
Comment (cviecco) [inline api.go]: return a 500 error here (internal server error)

File: api.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+//All handlers and API endpoints starts from here.
+
+//Display all groups in Target LDAP --required
+func (state *RuntimeState) GetallGroups_Handler(w http.ResponseWriter, r *http.Request) {
+	var AllGroups_TargetLdap GetGroups
+
+	Allgroups, err := state.get_allGroups(state.Config.TargetLDAP.GroupSearchBaseDNs)
+
+	if err != nil {
+		log.Panic(err)
```
Comment (cviecco) [inline api.go]: remove the panic, log the error and return a 500

File: api.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+
+
+
+//Displays all Groups of a User. --required
+func (state *RuntimeState) GetGroupsofUser_Handler(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	params, ok := q["username"]
+	if ok {
```
Comment (cviecco) [inline api.go]: invert the logic here to have less branches:  if !ok {log and return} then continue with the valid code

File: api.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+		var user_groups GetUserGroups
+
+		user_groups.UserName = params[0] //username is "cn" Attribute of a User
+		UsersAllgroups, err := state.getGroupsOfUser(state.Config.TargetLDAP.GroupSearchBaseDNs, user_groups.UserName)
+		sort.Strings(UsersAllgroups)
+		user_groups.UserGroups = UsersAllgroups
+
+		if err != nil {
+			log.Println(err)
```
Comment (cviecco) [inline api.go]: and return a 500

File: api.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	http.NotFoundHandler()
+}
+
+
+
+//Displays All Users in a Group --required
+func (state *RuntimeState) GetUsersinGroup_Handler(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	params, ok := q["groupname"]
+	if ok {
```
Comment (cviecco) [inline api.go]: Same as previous function. if !ok return and on all the errors, log it (without a panic) and then return a 500

File: database.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+//Initialsing database
+func initDB(state *RuntimeState) (err error) {
+
+	state.dbType = "sqlite3"
+	state.db, err = sql.Open("sqlite3", "./ldap-group-management1.db")
+	if err != nil {
+		return err
+	}
+	if true {
+		sqlStmt := `create table if not exists pending_requests (username text not null, groupname text not null, time_stamp int not null);`
```
Comment (cviecco) [inline database.go]: also add a unique index on username + groupname (let the db help you enforce constraints)

File: email.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+	// Set the sender and recipient.
+	c.Mail("ldap-group-manager-noreply@example.net")
+	for _, recipient := range users_email {
+		//c.Rcpt("recipient@example.net")
+		c.Rcpt(recipient)
+	}
+	// Send the email body.
+	wc, err := c.Data()
+	if err != nil {
+		log.Fatal(err)
```
Comment (cviecco) [inline email.go]: this is non-fatal, just return the err

File: email.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
```
Comment (cviecco) [inline email.go]: this does not need to be fatal

File: handlers.go [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+	"net/http"
+	"log"
+	"sort"
+	"fmt"
+	"strings"
+	"encoding/json"
+)
+
+//Main page with all LDAP groups displayed
+func  (state *RuntimeState) Index_Handler(w http.ResponseWriter, r *http.Request) {
```
Comment (cviecco) [inline handlers.go]: use golang NamingConvetions for functions (CamelCase or camelCase)

File: handlers.go [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+	userInfo, err := authSource.GetRemoteUserInfo(r)
+	if err != nil {
+		panic(err)
```
Comment (cviecco) [inline handlers.go]: in all handlers: dont panic on error, but log and return a 500 (internal server error)

File: api.go (hunk 2 above)
Comment (puneethreddy20) [reply api.go]: returned the error!

File: handlers.go [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+				response.Pending_actions=append(response.Pending_actions,entry)
+			}else{
+				continue
+			}
+		} else if state.isGroupMemberorNot(description,username){
+			response.Pending_actions=append(response.Pending_actions,entry)
+		}
+		continue
+	}
+	if response.Pending_actions==nil{
```
Comment (cviecco) [inline handlers.go]: changed to something similar to my slack messge

File: handlers.go [hunk 13]
```diff
@@ -0,0 +1,572 @@
...
+		return
+	}
+	if userInfo == nil {
+		log.Println("null userinfo!")
+		http.Error(w,"null userinfo",http.StatusInternalServerError)
+		return
+	}
+	//vals:=r.URL.Query()
+	username:=*userInfo.Username
+	if state.userisAdminOrNot(username){
```
Comment (cviecco) [inline handlers.go]: change to early exit

File: handlers.go [hunk 14]
```diff
@@ -0,0 +1,572 @@
...
+		return
+	}
+	if userInfo == nil {
+		log.Println("null userinfo!")
+		http.Error(w,"null userinfo",http.StatusInternalServerError)
+		return
+	}
+	//vals:=r.URL.Query()
+	username:=*userInfo.Username
+	if state.userisAdminOrNot(username) {
```
Comment (cviecco) [inline handlers.go]: same as above, change to early exit

File: email.go [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+		target_address=nil
+	}
+	return nil
+}
+
+
+func (state *RuntimeState) RejectRequestemail(requesteduser string,otheruser string,users_email []string,
+	groupname string,remoteAddr string,userAgent string) error{
+	// Connect to the remote SMTP server.
+	c, err := smtp.Dial("smtp.example.net:25")
```
Comment (cviecco) [inline email.go]: smtp server should be a configuration option

File: email.go [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+const requestaccess_mailTemplateText = `Subject: Requesting access for group {{.Groupname}}
+
+User {{.RequestedUser}} has requested access for group {{.Groupname}} (from {{.RemoteAddr}} ({{.Browser}} {{.OS}} ))`
+
+//send email for requesting access to a group
+func (state *RuntimeState) SuccessRequestemail(requesteduser string,users_email []string,
+	groupname string,remoteAddr string,userAgent string) error{
+	// Connect to the remote SMTP server.
+	c, err := smtp.Dial("smtp.example.net:25")
```
Comment (cviecco) [inline email.go]: smtp server from configuration

File: email.go [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+		target_address=nil
+	}
+	return nil
+}
+
+//for approving requests in pending actions main email function
+func (state *RuntimeState) approveRequestemail(requesteduser string,otheruser string,users_email []string,
+	groupname string,remoteAddr string,userAgent string) error{
+	// Connect to the remote SMTP server.
+	c, err := smtp.Dial("smtp.example.net:25")
```
Comment (cviecco) [inline email.go]: smtp server from configuration

File: handlers.go [hunk 18]
```diff
@@ -0,0 +1,564 @@
...
+	"log"
+	"sort"
+	"fmt"
+	"strings"
+	"encoding/json"
+)
+
+//Main page with all LDAP groups displayed
+func  (state *RuntimeState) IndexHandler(w http.ResponseWriter, r *http.Request) {
+	userInfo, err := authSource.GetRemoteUserInfo(r)
```
Comment (cviecco) [inline handlers.go]: as discussed wrap this into a single function that gets you the username

**Pull request merged**